File: src/main/java/frc/robot/subsystems/launcher/LauncherIO.java
package frc.robot.subsystems.launcher;

import org.littletonrobotics.junction.AutoLog;

import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Current;
import edu.wpi.first.units.measure.Temperature;
import edu.wpi.first.units.measure.Voltage;

import static edu.wpi.first.units.Units.*;

public interface LauncherIO {
    
    public default void updateInputs(LauncherIOInputs input){}
    
    public default void setGoal(AngularVelocity goal){}
    public default void setVout(Voltage vout){}

    @AutoLog
    public class LauncherIOInputs{
        public Angle angle = Rotations.zero();
        public AngularVelocity speed = RPM.zero();

        public Voltage motorVoltageOut = Volts.zero();
        public Current motorCurrentOut = Amps.zero();
        public Temperature motorTemp = Celsius.zero();

        public AngularVelocity goal = RPM.zero();
        public boolean atSetpoint = false;

        public boolean openLoop = false;

    }
}

File: src/main/kotlin/frc/robot/subsystems/launcher/LauncherIO.kt
package frc.robot.subsystems.launcher

import edu.wpi.first.units.Units.Amps
import edu.wpi.first.units.Units.Celsius
import edu.wpi.first.units.Units.RPM
import edu.wpi.first.units.Units.Rotations
import edu.wpi.first.units.Units.Volts
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.AngularVelocity
import edu.wpi.first.units.measure.Current
import edu.wpi.first.units.measure.Temperature
import edu.wpi.first.units.measure.Voltage
import org.littletonrobotics.junction.AutoLog

interface LauncherIO {
    fun updateInputs(input: LauncherIOInputs) {}

    fun setGoal(goal: AngularVelocity) {}
    fun setVout(vout: Voltage) {}

    @AutoLog
    open class LauncherIOInputs {
        @JvmField var angle: Angle = Rotations.zero()
        @JvmField var speed: AngularVelocity = RPM.zero()

        @JvmField var motorVoltageOut: Voltage = Volts.zero()
        @JvmField var motorCurrentOut: Current = Amps.zero()
        @JvmField var motorTemp: Temperature = Celsius.zero()

        @JvmField var goal: AngularVelocity = RPM.zero()
        @JvmField var atSetpoint: Boolean = false

        @JvmField var openLoop: Boolean = false
    }
}
